### ISSUE TITLE
imsave gives blank pdf/eps 

### ISSUE THREAD

**Issue by FlashTek** [matplotlib/matplotlib#9203]
### Bug report

**Bug summary**

the `plt.imsave(..)` method creates a blank file using the `.pdf/.eps` format.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import numpy as np
data = np.arange(100*100).reshape(100, 100).astype(float)
plt.imsave("test.pdf", data, origin="lower", format="pdf") #same behavior also with eps instead of pdf
```

**Actual outcome**

A blank `.pdf/.eps` is created (see [test.pdf](https://github.com/matplotlib/matplotlib/files/1313908/test.pdf))

**Expected outcome**
A non-blank `.pdf` which shows the same plot as the `.png` version of the code:
[![][1]][1]

**Matplotlib version**

  * Operating System: Windows 10 Pro x64
  * Matplotlib Version: 2.0.2
  * Python Version: 3.6
  * Other Libraries: numpy 1.12.1

I installed `matplotlib` via `anaconda`.

  [1]: https://i.stack.imgur.com/3j6D7.png

**Comment by dstansby** [comment]
I can reproduce this on the `master` branch and OSX.

**Comment by ImportanceOfBeingErnest** [comment]
Maybe it is worth noting that the output is different for different backends, while still being completely undesired. Using python 2.7 on windows 8.1 (64bit) and matplotlib 2.0.2, 

* with the `Agg` or `pdf` backend, the output is a completely white pdf
![image](https://user-images.githubusercontent.com/23121882/30589290-357ca77c-9d3a-11e7-8498-ffa044d32d53.png)

* while with `Qt4Agg` the output shows exactly one small pixel in the lower left corner
![image](https://user-images.githubusercontent.com/23121882/30589426-a48f3f08-9d3a-11e7-983b-f58e06679271.png)

**Comment by tacaswell** [comment]
This is a pretty major regression, I am inclined to block 2.1 on this.

`imsave` uses `Figure.figimage` and `mimage.FigureImage` which I am not sure are used anywhere else.  I have a vague memory of fixing a similar bug for `imshow` (which uses `AxesImage`) that probably just needs to be applied here.

**Comment by jklymak** [comment]
So, `FigureImage` clips the image to the size of the figure, but was using `renderer.width` and `renderer.height`. If I change this to 

```python
width, height = renderer.get_canvas_width_height()
        clip = Bbox([[0, 0], [width, height]])

**Comment by jklymak** [comment]
Possible fix?  https://github.com/matplotlib/matplotlib/pull/9204

**Comment by jklymak** [comment]
Ooops, no.  The `pdf` canvas is hardwired at dpi=72 for some reason, so `renderer.get_canvas_width_height()` doesn't gove the "correct" width/height either in the user-defined `dpi`.  So relying on the pdf canvas size doesn't work for the clipping.

Frustratingly, for `renderer.width`,`RenderAgg` returns `fig.width * fig.dpi`

For `MixedModeRenderer`, `renderer.width` simply returns `fig.width`.  This is the root of the bug since `clip` in the above is in pixels.  I don't know how to fix the root problem without breaking lots of other things.

**Comment by WeatherGod** [comment]
The pdf standard dictates that dpi is 72.

**Comment by jklymak** [comment]
Not sure what you mean.  I can place a 200 dpi raster into a pdf w/o a problem, can't I?

**Comment by dopplershift** [comment]
@tacaswell If it's broken on 2.x already, I'm not sure you *need* to block 2.1 on this. That said, I think it depends heavily on how long it takes to fix. I wouldn't delay 2.1 past the end of the month for a bug that already exists in several releases.

**Comment by WeatherGod** [comment]
as I understand it, pdf just has a dpi of 72 just as a reference point for
computing the extent of the page sizes and the extent of any rasterized
components. Everything is resampled on-the-fly when displayed, so the
particular dpi itself is irrelevant, so long as the writer and the reader
both aggree on the same dpi value.

**Comment by jklymak** [comment]
OK, my apologies, maybe I'm not being clear because we don't have the right example...

I got onto this by trying to make the tests work.  I think the example above works fine w/ my previous fix.  However in general this example from the tests clips the 200x200 pixel image off at 144x144 if I render in pdf...

```python

import matplotlib
matplotlib.use('pdf')
import matplotlib.pyplot as plt
import numpy as np
for suppressComposite in [False]:
    fig = plt.figure(figsize=(2,2), dpi=100)
    fig.suppressComposite = suppressComposite
    x,y = np.ix_(np.arange(100.0)/100.0, np.arange(100.0)/100.0)
    z = np.sin(x**2 + y**2 - x*y)
    c = np.sin(20*x**2 + 50*y**2)
    img = z + c/5

    fig.figimage(img, xo=0, yo=0, origin='lower')
    fig.figimage(img[::-1,:], xo=0, yo=100, origin='lower')
    fig.figimage(img[:,::-1], xo=100, yo=0, origin='lower')
    fig.figimage(img[::-1,::-1], xo=100, yo=100, origin='lower')

    if suppressComposite:
        fig.savefig('Boo.pdf', dpi=100)
        fig.savefig('Boo.png', dpi=100)
    else:
        fig.savefig('Boo2.pdf', dpi=100)
        #fig.savefig('Boo2.png', dpi=100)

```

What happens is that `figimage` ignores the `dpi` call in `figure` and clips the figure to 72dpi when the `MixedModeRenderer` (pdf) is used.  

But maybe thats OK?  The docs for `figimage` say:

```
figimage complements the axes image
        (:meth:`~matplotlib.axes.Axes.imshow`) which will be resampled
        to fit the current axes.  If you want a resampled image to
        fill the entire figure, you can define an
        :class:`~matplotlib.axes.Axes` with extent [0,0,1,1].
```

...so does the PDF backend simply always resample and `fig.figimage` is poorly defined for `pdf`?  I guess I'd be OK w/ that logic.

**Comment by FlashTek** [comment]
@dopplershift If the `2.1` update will not be blocked until this issue is solved, I think it would at least be useful to show a warning when using the function for `pdf` files. Otherwise, a user can spend a long time in which he doesn't know, if he's doing something wrong, or `matplotlib` is not working properly.

**Comment by WeatherGod** [comment]
I think that using the "dpi" keyword for savefig() of a pdf (and I think
eps/ps) should probably be considered an error for anything but 72. But,
I'll admit, I haven't considered the implications.

**Comment by jklymak** [comment]
@WeatherGod   `Test.pdf` and `Test2.pdf` look very different!  FWIW, this is *the* number-one reason I switched to matplotlib!

```python

fig, ax = plt.subplots(figsize=(2,2))
ax.pcolormesh(np.random.rand(100,100), rasterized=True)
fig.savefig('Test.pdf', dpi=300)
fig.savefig('Test2.pdf', dpi=30)
```

[Test.pdf](https://github.com/matplotlib/matplotlib/files/1315894/Test.pdf)
[Test2.pdf](https://github.com/matplotlib/matplotlib/files/1315893/Test2.pdf)

**Comment by jklymak** [comment]
OK, re-opened #9204.  New approach seems OK.

I will still argue that hardwiring the pdf backend to have dpi=72 is not practically useful, and makes it hard to deal with rasterized objects which need not have pixels-per-inch=72, which is how `dpi` is used in savefig (appropriately in my opinion).

**Comment by jklymak** [comment]
The PDF spec makes lots of reference to 1/72" as the default size of `UserUnit`, but I don't think that has anything fundamentally to do with *pixels-per-inch* of rasterized objects.  

To me, the `fig.savefig(fname, dpi=100., fortmat='pdf')` has always meant save any rasterized components at 100 *pixels-per-inch*.  Which it does just fine as is.  But, I'm not familiar enough with the backend machinery to understand why it needs to set the figure dpi to 72.

**Comment by tacaswell** [comment]
The reason that all of the the render has the dpi set to 72 is that it gets fed into the transform stack that converts data space coordinates to 'screen' space coordinates.  In the case of Agg we then feed those coordinates to the resterizer which colors the pixels, in the case of pdf we just write the values out.

Individual artists have a 'rasterized' flag that determines if they get output to the vector backends as vector data or if it should be rasterized and then embedded into the vector output as an image.  This can be useful to reduce ex 1M points in a scatter to something more manageable in the output.

**Comment by jklymak** [comment]
Right, but what is confusing is that I can do 

```python
matplotlib.use('pdf')

fig, ax = plt.subplots(figsize=(2, 2), dpi=200)
ax.pcolormesh( np.random.rand(400,400), rasterized=True)
fig.savefig('Boo.pdf', dpi=300)
```

and have the figure return different values for dpi depending on where I query `fig.dpi`.  I guess I can see why the dpi gets set by the pdf canvas, so that the figure transform gives the right distances, but it means that the original dpi=200 gets ignored.  As a user, I *expect* that this dpi will get carried through to the `savefig` statement so that I don't need to set the dpi at that stage.  For instance:

```python

matplotlib.use('pdf')

fig, ax = plt.subplots(figsize=(2,2), dpi=10)
print(fig.dpi)
ax.pcolormesh( np.random.rand(400,400), rasterized=True)
fig.savefig('Boo10.pdf')
```

completely ignores `dpi=10` and saves the raster at `dpi=100` (my default).  But the print statement returns `10`.  

Even more surprising, if I do `savefig('Boo10.png')` it still ignores the dpi=10.  I find this pretty confusing, but maybe its fine.

**Comment by tacaswell** [comment]
If you pass 'dpi='figure'' to save fig (or set it in you rcparams) the dpi on the figure will be respected in savefig.  As of 2.0 that is the default (via 016406acccd665796a0a5dbc6448912ea450400c).  

The reason for having a separate configuration for the savefig dpi (as far as I know) is lost to the sands of time.  Changing the default to `'figure'` is the next best thing to deprecating it.

**Comment by jklymak** [comment]
`dpi='figure'` is certainly not the default on master using the PDF backend.  Or rather, perhaps it is saving at 72 dpi, which is what the pdf backend sets the figure dpi to, but its certainly not respecting the `dpi=10` that I used in the `subplots` call.  

```python

import matplotlib
matplotlib.use('pdf')
import matplotlib.pyplot as plt
import numpy as np

X = np.random.rand(400,400)

fig, ax = plt.subplots(figsize=(2,2), dpi=10)
print(fig.dpi)
ax.pcolormesh(X, rasterized=True)
fig.savefig('Boo10.pdf')

fig, ax = plt.subplots(figsize=(2,2), dpi=72)
print(fig.dpi)
ax.pcolormesh(X, rasterized=True)
fig.savefig('Boo72.pdf')
```

yield exactly the same files....  `diff Boo10.pdf Boo72.pdf` returns no differences.

[Boo10.pdf](https://github.com/matplotlib/matplotlib/files/1318343/Boo10.pdf)
[Boo72.pdf](https://github.com/matplotlib/matplotlib/files/1318345/Boo72.pdf)

**Comment by tacaswell** [comment]
Closed by #9204
